### PR TITLE
Switch to the `hatchling` build backend.

### PR DESCRIPTION
**Pull request by jsirois**
This should elminate flaky CI due to ephemeral dirs in the source root
created by the setuptools build backend which were not multi-process
safe.

The `hatchling` build backend supports metadata plugins and it performs
isolated builds making it a good replacement for the prior setuptools
hack for effecting dynamic `Requires-Python` to enable Python 3.13
shadow-support. Unfortunately, there is no nice way to specify an
in-tree plugin; so we actually use a very thin in-tree build backend
that wraps `hatchling.build`, to allow discovery of our in-tree plugin.

### CODE REVIEW

File: pex/pip/tool.py [hunk 1]
```diff
@@ -451,7 +452,7 @@ def spawn_download_distributions(
 
         if not build_isolation:
             download_cmd.append("--no-build-isolation")
-            extra_env.update(PEP517_BACKEND_PATH=os.pathsep.join(sys.path))
+            pex_extra_sys_path.extend(sys.path)
```
Comment (jsirois) [inline pex/pip/tool.py:455]: Now that Pex uses an in-tree build backend, it unearthed this latent bug where `PEP517_BACKEND_PATH` would get stomped with the backend-path. As such, both here in spawn download and lower in spawn install wheel, we now pass non-isolated-build sys.path via PEX_EXTRA_SYS_PATH instead.

File: scripts/package.py [hunk 2]
```diff
@@ -46,7 +47,12 @@ def build_pex_pex(output_file: PurePath, verbosity: int = 0) -> None:
     subprocess.run(args, check=True)
 
 
-def describe_git_rev() -> str:
+def describe_rev() -> str:
+    if not os.path.isdir(".git") and os.path.isfile("PKG-INFO"):
```
Comment (jsirois) [inline scripts/package.py:51]: This was always a bug, I just never tried unpacking our sdist and doing a build from there before. That now works.

File: tests/build_system/test_pep_518.py [hunk 3]
```diff
@@ -68,14 +68,15 @@ def test_load_build_system_pyproject(
...
     dists = {
         dist.metadata.project_name
         for dist in Virtualenv(build_system.venv_pex.venv_dir).iter_distributions()
     }
-    assert ProjectName("setuptools") in dists
+    assert ProjectName("hatchling") in dists
     subprocess.check_call(
-        args=[build_system.venv_pex.pex, "-c", "import {}".format(build_system.build_backend)]
+        args=[build_system.venv_pex.pex, "-c", "import {}".format(build_system.build_backend)],
+        env=build_system.env,
```
Comment (jsirois) [inline tests/build_system/test_pep_518.py:79]: This env carries the backend-path -> PEX_EXTRA_SYS_PATH which is needed to have the build see our now in-tree build backend.

File: tests/integration/build_system/test_pep_518.py [hunk 4]
```diff
@@ -55,5 +55,6 @@ def test_load_build_system_pyproject_custom_repos(
     build_system = load_build_system(current_target, custom_resolver, pex_project_dir)
     assert isinstance(build_system, BuildSystem)
     subprocess.check_call(
-        args=[build_system.venv_pex.pex, "-c", "import {}".format(build_system.build_backend)]
+        args=[build_system.venv_pex.pex, "-c", "import {}".format(build_system.build_backend)],
+        env=build_system.env,
```
Comment (jsirois) [inline tests/integration/build_system/test_pep_518.py:59]: Ditto https://github.com/pantsbuild/pex/pull/2331/files#r1455083605

**Review by kaos — APPROVED** [review]

**Review by benjyw — COMMENTED** [review]
Late to the party, but LGTM

**Pull request merged**